core-ktx/src/main/java/com/amplifyframework/kotlin/Auth.kt

```diff
@@ -0,0 +1,269 @@
+     * @param options Advanced options such as additional attributes of the user
+     *                or validation data.
+     *                If not provided, default options will be used.
+     * @return A sign-up result; check the nextStep field to determine which
+     *         action to take next. The next step is typically to confirm a
+     *         code sent over email/SMS.
+     */
+    @Throws(AuthException::class)
+    suspend fun signUp(
+        username: String,
+        password: String,
+        options: AuthSignUpOptions = AuthSignUpOptions.builder().build()
+    ):
+        AuthSignUpResult
+
+    /**
+     * If you have attribute confirmation enabled, this will allow the user
+     * to enter the confirmation code they received to activate their account.
+     * @param username A login identifier e.g. `tony44`; or an email/phone number,
+     *                 depending on configuration
+     * @param confirmationCode The confirmation code the user received
+     * @return A sign-up result; if the code has been confirmed successfully, the result
+     *         will show true for isSignUpComplete().
+     */
+    @Throws(AuthException::class)
+    suspend fun confirmSignUp(username: String, confirmationCode: String): AuthSignUpResult
+
+    /**
+     * If the user's code expires or they just missed it, this method can
+     * be used to send them a new one.
+     * @param username A login identifier e.g. `tony44`; or an email/phone number,
+     *                 depending on configuration
+     * @return A sign-up result; if the code is requested, typically the result will
+     *         include a next step requiring confirmation of the re-sent code.
+     */
+    @Throws(AuthException::class)
+    suspend fun resendSignUpCode(username: String): AuthSignUpResult
+
+    /**
+     * Basic authentication to the app with a username and password or, if custom auth is setup,
+     * you can send null for those and the necessary authentication details in the options object.
+     * @param username A login identifier e.g. `tony44`; or an email/phone number, depending on configuration.
+     *                 May be omitted or null when using custom auth.
+     * @param password User's password for normal sign-up. May be omitted or null if custom auth or
+     *                 password-less configurations are in use
+     * @param options Advanced options such as a map of auth information for custom auth,
+     *                If not provided, default options will be used
+     * @return A sign-in result. The nextStep field may indicate additional actions to be taken
+     *         to confirm the sign-in, or it may show isSignInComplete as true, in which case
+     *         an authenticated session is available
+     */
+    @Throws(AuthException::class)
+    suspend fun signIn(
+        username: String? = null,
+        password: String? = null,
+        options: AuthSignInOptions? = null
```

Comment:
Yes! Good eye. Most of the other APIs in this PR have some kind of default object that I can use as a default parameter.

[`AuthSignInOptions` is an abstract class](https://github.com/aws-amplify/amplify-android/blob/72b5820ff217f5ca6cde715aeb2ece523f79b274/core/src/main/java/com/amplifyframework/auth/options/AuthSignInOptions.java#L24). As is right now, I can't instantiate it, so I sued `null` instead.

One alternative would be to add a factory method to that abstract class, like:
```kotlin
AuthSignInOptions.defaults()
```
It would create a new sentinel object, `DefaultAuthSignInOptions`. We could look for that type inside our implementation and treat it specially.

On second thought, this does seem like the better approach. I can update the code base to add this mechanism, if it sounds good to you.

Reply:
> I can update the code base to add this mechanism, if it sounds good to you.

Sounds reasonable.  I think the current approach is fine too.  Either way seems fine to me!